### PR TITLE
perf(multimodal): avoid feature hashing for pad values

### LINKED ISSUE #1029: [Feature] Optimize multimodal pad-value hashing for VLM requests
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sgl-jax/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

### Motivation

Single-image VLM requests currently derive `MultimodalDataItem.pad_value` by hashing the processed feature tensor when no precomputed hash is present. In a v6e-4 profile of `Qwen/Qwen2.5-VL-7B-Instruct`, this tokenizer-side feature hashing showed up as a measurable host-side request cost:

| Event | Calls | Mean | Total |
|---|---:|---:|---:|
| `MultimodalDataItem.set_pad_value` | 65 | `3412.899us` | `221.838ms` |
| `hash_feature` | 65 | `3316.494us` | `215.572ms` |

The tokenizer already has access to the raw image/video/audio payload bytes before feature preprocessing. It should be able to carry a stable content hash into `MultimodalDataItem.hash`, allowing `set_pad_value()` to avoid walking large processed feature tensors on the request path.

### Proposed change

- Compute or reuse a stable raw payload/content hash while loading image, video, and audio payloads in `multimodal_tokenizer.py`.
- Set `MultimodalDataItem.hash` before `set_pad_value()` runs.
- Keep `hash_feature` as the fallback for callers without a precomputed hash.
- Include modality and relevant preprocessing metadata in the hash where needed so cache keys remain distinct.

### Validation plan

Unit tests:
- Precomputed hash skips feature hashing.
- Fallback still calls `hash_feature`.
- `pad_input_tokens` uses the precomputed hash.
- Dict round-trip preserves hash and pad-value behavior.
- Image payload hashing is stable and content-sensitive.

Profiling confirmation on v6e-4 with `Qwen/Qwen2.5-VL-7B-Instruct`:

| Metric | Baseline: feature-tensor hash | Optimized: payload hash |
|---|---:|---:|
| Measured requests | `64/64` HTTP 200 | `64/64` HTTP 200 |
| `hash_feature` calls | `65` | `0` |
| `hash_feature` mean | `3316.494us` | N/A |
| `hash_feature` total | `215.572ms` | N/A |
| `MultimodalDataItem.set_pad_value` calls | `65` | `65` |
| `MultimodalDataItem.set_pad_value` mean | `3412.899us` | `31.320us` |
| `MultimodalDataItem.set_pad_value` total | `221.838ms` | `2.036ms` |
| Request latency median | `0.3174s` | `0.3115s` |
| Request latency p95 | `0.3247s` | `0.3180s` |

### Related resources

- `python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py`
- `python/sgl_jax/srt/multimodal/common/modality_enum.py`
- Related inspiration: Modal's multimodal inference optimization write-up discusses using stable media-content hashes to avoid repeatedly hashing processed tensors on the request path: https://modal.com/blog/boosting-multimodal-inference-performance-by-greater-than-10-with-a-single-python-dictionary

### PR DESCRIPTION
**Pull request by hashkanna**
## Motivation

Fixes #1029.

Single-image VLM requests currently derive `MultimodalDataItem.pad_value` by hashing the processed feature tensor when no precomputed hash is present. In a v6e-4 profile of `Qwen/Qwen2.5-VL-7B-Instruct`, this tokenizer-side feature hashing showed up as a measurable host-side request cost:

| Event | Calls | Mean | Total |
|---|---:|---:|---:|
| `MultimodalDataItem.set_pad_value` | 65 | `3412.899us` | `221.838ms` |
| `hash_feature` | 65 | `3316.494us` | `215.572ms` |

The tokenizer already has access to the raw image/video/audio payload bytes before feature preprocessing. This PR carries a stable content hash into `MultimodalDataItem.hash`, allowing `set_pad_value()` to avoid walking large processed feature tensors on the request path.

Related inspiration: Modal's multimodal inference optimization write-up discusses the same general principle of carrying stable media-content identity through preprocessing instead of repeatedly hashing large processed tensors on the request path: https://modal.com/blog/boosting-multimodal-inference-performance-by-greater-than-10-with-a-single-python-dictionary

## Modifications

- Add an internal loaded-payload wrapper that carries both decoded multimodal data and a stable payload hash.
- Compute payload hashes for image, video, and audio inputs while raw bytes are available.
- Include modality and relevant preprocessing metadata in hashes where needed so cache keys remain distinct.
- Set `MultimodalDataItem.hash` before `set_pad_value()` runs.
- Keep `hash_feature` as the fallback for callers without a precomputed hash.
- Add focused unit tests for precomputed hash behavior, fallback behavior, dict round-trip behavior, and image payload hash stability.

## Accuracy Tests

No model output changes are expected. This changes the way multimodal pad values are derived for cache-key differentiation, while preserving the existing feature-hash fallback.

Local validation:

- [x] `uv run --project python python -m unittest sgl_jax.test.multimodal.test_multimodal_pad_value_hash`
- [x] `python3 -m py_compile python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py python/sgl_jax/test/multimodal/test_multimodal_pad_value_hash.py`
- [x] `ruff check python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py python/sgl_jax/test/multimodal/test_multimodal_pad_value_hash.py`
- [x] `pre-commit run --all-files --show-diff-on-failure`

TPU validation:

- [x] Focused unittest passed on v6e-4 before the profiling confirmation run.

## Benchmarking and Profiling

Profiling confirmation on v6e-4 with `Qwen/Qwen2.5-VL-7B-Instruct`:

| Metric | Baseline: feature-tensor hash | Optimized: payload hash |
|---|---:|---:|
| Measured requests | `64/64` HTTP 200 | `64/64` HTTP 200 |
| `hash_feature` calls | `65` | `0` |
| `hash_feature` mean | `3316.494us` | N/A |
| `hash_feature` total | `215.572ms` | N/A |
| `MultimodalDataItem.set_pad_value` calls | `65` | `65` |
| `MultimodalDataItem.set_pad_value` mean | `3412.899us` | `31.320us` |
| `MultimodalDataItem.set_pad_value` total | `221.838ms` | `2.036ms` |
| Request latency median | `0.3174s` | `0.3115s` |
| Request latency p95 | `0.3247s` | `0.3180s` |

The optimized run had one request-level outlier at `8.32436s`; median and p95 remained stable.

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update. N/A.

### CODE REVIEW

File: python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py [hunk 1]
```diff
@@ -748,89 +811,178 @@ def _resize_video_frames(self, video_np: np.ndarray, video_config: dict) -> np.n
...
             source = source.url
         if isinstance(source, bytes):
-            return Image.open(io.BytesIO(source)).convert("RGB")
+            return _LoadedMultimodalPayload(
+                data=Image.open(io.BytesIO(source)).convert("RGB"),
+                hash=self._hash_payload(source, "image"),
+            )
         if os.path.exists(source):
-            return Image.open(source).convert("RGB")
+            with Image.open(source) as image:
```
Comment (pengchengneo) [inline python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py]: the code decodes the media from the path first, then reopens the same path to compute the hash. If the file changes between those two reads, the feature and hash may come from different contents, producing an incorrect radix cache key. 

Please read the file bytes once and use the same bytes for both decoding and hashing.

File: python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py [hunk 2]
```diff
@@ -748,89 +811,178 @@ def _resize_video_frames(self, video_np: np.ndarray, video_config: dict) -> np.n
...
+        return self._hash_payload(
+            array.tobytes(),
+            *metadata,
+            str(array.dtype),
+            tuple(int(dim) for dim in array.shape),
+        )
+
+    def _combine_mm_hashes(self, hashes: list[int | None], modality: str) -> int | None:
+        valid_hashes = [value for value in hashes if value is not None]
+        if not valid_hashes:
```
Comment (pengchengneo) [inline python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py]: _combine_mm_hashes() silently drops None hashes. If any payload is processed successfully but its hash is unavailable, the combined hash may only cover part of the multimodal input instead of falling back to feature hashing. 

I think this should return None whenever any input hash is None, so set_pad_value() can use the existing feature-hash fallback.

File: python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py (hunk 1 above)
Comment (hashkanna) [reply python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py]: thanks, fixed it in commit `9b2f20dd`. Local file inputs now create a temporary snapshot while streaming the source file in chunks. The hash is computed over the exact bytes copied to that snapshot, and image/video/audio/Qwen-video decoding reads from the same snapshot.

File: python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py (hunk 2 above)
Comment (hashkanna) [reply python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py]: `_combine_mm_hashes()` now returns `None` if any input hash is unavailable, so `MultimodalDataItem.set_pad_value()` falls back to the existing feature hashing path instead of building a partial payload hash. Also, added a focused unit test for this case.

File: python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py (hunk 1 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/multimodal/manager/multimodal_tokenizer.py]: ok

**Comment by pengchengneo** [comment]
LGTM

**Review by pengchengneo — DISMISSED** [review]

**Comment by pathfinder-pf** [comment]
@hashkanna please fix ci error

**Comment by hashkanna** [comment]
> @hashkanna please fix ci error

@pathfinder-pf The failing job is the known flake in test_bench_serving_dense_tp_4 that #1077 just fixed. Rebasing onto main to pick up the fix.

**Comment by hashkanna** [comment]
@pengchengneo is there anything further I need to do here to get it merged?

**Review by pengchengneo — APPROVED** [review]
